Fix hooks for lazy modules: allow pre-forward & forward hook self-removal

ertrue opened the pull request:
Summary:
This is to fix the module support on lazymodulefixin on the bug issue #60132
Check the link: https://github.com/pytorch/pytorch/issues/60132

Test Plan: Unit test pass

Differential Revision: D29274068

jbschlosser commented on torch/nn/modules/module.py:
```diff
@@ -1059,9 +1059,9 @@ def _call_impl(self, *input, **kwargs):
             full_backward_hooks, non_full_backward_hooks = self._get_backward_hooks()
         if _global_forward_pre_hooks or self._forward_pre_hooks:
-            for hook in itertools.chain(
+            for hook in list(itertools.chain(
```
Instead of using `list()` + `itertools.chain()`, can you please try:
```python
for hook in (*_global_forward_pre_hooks.values(), *self._forward_pre_hooks.values()):
```

jbschlosser commented on test/test_nn.py:
```diff
@@ -16536,6 +16536,44 @@ def forward_hook(m, input, output):
+        handle = module.register_forward_pre_hook(removable_hook)
+        handle_2 = module.register_forward_pre_hook(removable_hook_2)
+        input = torch.randn(2, 2)
+        output = module(input)
```
Might want to verify that the hooks have actually been removed after the forward pass

jbschlosser commented on test/test_nn.py:
```diff
@@ -16536,6 +16536,44 @@ def forward_hook(m, input, output):
+            handle_2.remove()
+            return input
+
+        handle = module.register_forward_pre_hook(removable_hook)
+        handle_2 = module.register_forward_pre_hook(removable_hook_2)
```
Should these lines reference `register_forward_hook()` instead?

jbschlosser commented on test/test_nn.py:
```diff
@@ -16536,6 +16536,44 @@ def forward_hook(m, input, output):
         expected_grad = -sig_x * (1 - sig_x) * 2 * mask
         self.assertEqual(input.grad, expected_grad)
 
+    def test_module_forward_preforward_hook_removable(self):
```
Could you please add a comment or two for these tests describing what they're testing for posterity?

jbschlosser reviewed (commented):
Thanks for the fix! Couple minor comments below

albanD reviewed (approved):
The updated code looks good to me (maybe a comment in the doc would help).

@jbschlosser all good for you as well?

jbschlosser reviewed (approved):
LGTM too! Would never hurt to have more comments of course :)

Commit:
[torch] Module fix on the support of LazyModule on bug #60132 (#60517)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/60517

This is to fix the module support on lazymodulefixin on the bug issue #60132
Check the link: https://github.com/pytorch/pytorch/issues/60132

We will have to update lazy_extension given the dependency on module.py and update the unit test as well.

Test Plan:
Unit test passes

torchrec test passes

Reviewed By: albanD

Differential Revision: D29274068

fbshipit-source-id: eb8a45976b28fe2be5df74d5fc52370975d6ec99
